File: app/src/main/java/com/compose/myapplication/SavePhoto.java
package com.compose.myapplication;

import android.app.Activity;
import android.content.Context;
import android.media.MediaScannerConnection;
import android.os.Environment;
import android.widget.Toast;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * The save photo class allows for the user to save a high resolution version of
 * the puzzle they just solved.
 *
 * @author Rick
 */
public class SavePhoto {

    CommonVariables commonVariables = CommonVariables.getInstance();
    int currentImageToSave;
    Toast toast = null;

    public SavePhoto(Context context, final int currentImage) {
        currentImageToSave = currentImage;

        final Activity activity = (Activity) context;
        activity.runOnUiThread(new Runnable() {

            private boolean mExternalStorageAvailable = false;
            private boolean mExternalStorageWriteable = false;

            public void showToast(String message) {
                // Create and show toast for save photo
                toast = Toast.makeText(activity, message, Toast.LENGTH_SHORT);
                toast.show();
            }

            @Override
            public void run() {
                // save current image to devices images folder
                String state = Environment.getExternalStorageState();
                // check if writing is an option
                if (Environment.MEDIA_MOUNTED.equals(state)) {
                    // We can read and write the media
                    mExternalStorageAvailable = mExternalStorageWriteable = true;
                } else if (Environment.MEDIA_MOUNTED_READ_ONLY.equals(state)) {
                    // We can only read the media
                    mExternalStorageAvailable = true;
                    mExternalStorageWriteable = false;
                } else {
                    // Something else is wrong. It may be one of many other
                    // states, but
                    // all we need
                    // to know is we can neither read nor write
                    mExternalStorageAvailable = mExternalStorageWriteable = false;
                }

                if (mExternalStorageAvailable && mExternalStorageWriteable) {
                    // then write picture to phone
                    File path = Environment
                            .getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES);

                    final String name = commonVariables.data.artworks[currentImageToSave].titleOfArtist + "_"
                            + commonVariables.data.artworks[currentImageToSave].titleOfArtwork + ".jpeg";

                    File file = new File(path, name);

                    // check for file in directory
                    if (file.exists()) {
                        showToast("Photo Exists Already!");
                    } else {
                        try {
                            boolean b1 = path.mkdirs();
                            boolean b2 = path.exists();
                            // Make sure the Pictures directory exists.
                            if (b1 || b2) {

                                // get file into input stream
                                InputStream is = activity.getResources().openRawResource(
                                        commonVariables.data.artworks[currentImageToSave].imageID);

                                OutputStream os = new FileOutputStream(file);
                                byte[] data = new byte[is.available()];
                                is.read(data);
                                os.write(data);
                                is.close();
                                os.close();

                                commonVariables.imagesSaved++;
                                MainActivity act = (MainActivity) activity;
                                act.updatePuzzleStats();

                                showToast("Image Saved!");

                                MediaScannerConnection
                                        .scanFile(
                                                context,
                                                new String[]{file.toString()},
                                                null,
                                                (path1, uri) -> {
                                                    // new image is in phone
                                                    // database now for use

                                                });
                            } else {
                                showToast("Could not make/access directory.");
                            }
                        } catch (IOException e) {
                            showToast("Error making/accessing directory.");
                        }
                    }
                } else {
                    showToast("Directory not available/writable.");
                }
            }
        });
    }
}

File: app/src/main/java/com/compose/myapplication/SavePhoto.kt
package com.compose.myapplication

import android.content.Context
import android.media.MediaScannerConnection
import android.net.Uri
import android.os.Environment
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.io.OutputStream

/**
 * The save photo class allows for the user to save a high resolution version of
 * the puzzle they just solved.
 *
 * @author Rick
 */
class SavePhoto(private var c: Context?, private var currentImageToSave: Int) {

    suspend fun run(): FirstFragment.Companion.PHOTO_RESULT = withContext(Dispatchers.IO) {
        var success: FirstFragment.Companion.PHOTO_RESULT

        try {

            val mExternalStorageAvailable: Boolean
            val mExternalStorageWriteable: Boolean

            // save current image to devices images folder
            val state = Environment.getExternalStorageState()
            // check if writing is an option
            if (Environment.MEDIA_MOUNTED == state) {
                // We can read and write the media
                mExternalStorageWriteable = true
                mExternalStorageAvailable = true
            } else if (Environment.MEDIA_MOUNTED_READ_ONLY == state) {
                // We can only read the media
                mExternalStorageAvailable = true
                mExternalStorageWriteable = false
            } else {
                // Something else is wrong. It may be one of many other
                // states, but
                // all we need
                // to know is we can neither read nor write
                mExternalStorageWriteable = false
                mExternalStorageAvailable = false
            }
            if (mExternalStorageAvailable && mExternalStorageWriteable) {
                // then write picture to phone
                val path = Environment
                    .getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES)
                val name =
                    (CommonVariables.data.artworks[currentImageToSave].titleOfArtist + "_"
                            + CommonVariables.data.artworks[currentImageToSave].titleOfArtwork + ".jpeg")
                val file = File(path, name)

                // check for file in directory
                if (file.exists()) {
                    success = FirstFragment.Companion.PHOTO_RESULT.EXISTS
                } else {
                    try {
                        val b1 = path.mkdirs()
                        val b2 = path.exists()
                        // Make sure the Pictures directory exists.
                        if (b1 || b2) {

                            // get file into input stream
                            val `is` = c?.resources?.openRawResource(
                                CommonVariables.data.artworks[currentImageToSave].imageID
                            )
                            val os: OutputStream = FileOutputStream(file)
                            val data = `is`?.let { ByteArray(it.available()) }
                            `is`?.read(data)
                            os.write(data)
                            `is`?.close()
                            withContext(Dispatchers.IO) {
                                os.close()
                            }
                            CommonVariables.imagesSaved++
                            val act = c as MainActivity
                            act.updatePuzzleStats()
                            MediaScannerConnection
                                .scanFile(
                                    c, arrayOf(file.toString()),
                                    null
                                ) { _: String?, _: Uri? -> }
                            success = FirstFragment.Companion.PHOTO_RESULT.SAVED
                        } else {
                            success = FirstFragment.Companion.PHOTO_RESULT.ERROR
                        }
                    } catch (e: IOException) {
                        success = FirstFragment.Companion.PHOTO_RESULT.ERROR
                    }
                }
            } else {
                success = FirstFragment.Companion.PHOTO_RESULT.ERROR
            }
        } catch (_: Exception) {
            success = FirstFragment.Companion.PHOTO_RESULT.ERROR
        }
        return@withContext success
    }
}